src/uct/tcp/tcp_iface.c

```diff
@@ -406,26 +422,51 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_iface_t, uct_md_h md, uct_worker_h worker,
                                             params->stats_root : NULL)
                               UCS_STATS_ARG(params->mode.device.dev_name));
 
+    /* Maximum IOV count allowed by user's configuration and system constraints */
+    max_iov = ucs_min(config->max_iov, ucs_socket_max_iov());
+
     ucs_strncpy_zero(self->if_name, params->mode.device.dev_name,
                      sizeof(self->if_name));
-    self->outstanding           = 0;
-    self->config.prefer_default = config->prefer_default;
-    self->config.max_poll       = config->max_poll;
-    self->sockopt.nodelay       = config->sockopt_nodelay;
-    self->sockopt.sndbuf        = config->sockopt_sndbuf;
-    self->sockopt.rcvbuf        = config->sockopt_rcvbuf;
+    self->config.tx_seg_size      = config->tx_seg_size +
+                                    sizeof(uct_tcp_am_hdr_t);
+    self->config.rx_seg_size      = config->rx_seg_size +
+                                    sizeof(uct_tcp_am_hdr_t);
+    self->outstanding             = 0;
+    /* Consider TCP protocol and user's AM headers that use 1st and 2nd IOVs */
+    self->config.zcopy.max_iov    = ((max_iov >
+                                      UCT_TCP_EP_AM_ZCOPY_SERVICE_IOV_COUNT) ?
+                                     (max_iov -
+                                      UCT_TCP_EP_AM_ZCOPY_SERVICE_IOV_COUNT) : 0);
+    /* Use a remaining part of TX segment for AM Zcopy header */
+    self->config.zcopy.hdr_offset = (sizeof(uct_tcp_ep_zcopy_ctx_t) +
+                                     sizeof(struct iovec) *
+                                     (self->config.zcopy.max_iov +
+                                      UCT_TCP_EP_AM_ZCOPY_SERVICE_IOV_COUNT));
+    if (self->config.zcopy.hdr_offset > self->config.tx_seg_size) {
```

Comment:
should we fail only if max_iov > 0? according to qeery, max_iov should disable zcopy..